cmds/fspinfo/main.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2017-2018 the LinuxBoot Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// fspinfo prints FSP header information.
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+
+	"github.com/linuxboot/fiano/pkg/fsp"
+	"github.com/linuxboot/fiano/pkg/uefi"
+)
+
+var (
+	flagJSON = flag.Bool("j", false, "Output as JSON")
+)
+
+func main() {
+	flag.Parse()
+	if flag.Arg(0) == "" {
+		log.Fatal("Error: missing file name")
+	}
+	data, err := ioutil.ReadFile(flag.Arg(0))
+	if err != nil {
+		log.Fatalf("Error: cannot read input file: %v", err)
+	}
+
+	fv, err := uefi.NewFirmwareVolume(data, 0, false)
+	if err != nil {
+		log.Fatalf("Error: cannot parse Firmware Volume: %v", err)
+	}
+	if len(fv.Files) < 1 {
+		log.Fatal("Error: Firmware Volume has no files")
+	}
+	file := fv.Files[0]
+	// FIXME why does FSPH start at +4?
+	hdr, err := fsp.FromBytes(file.Buf()[file.DataOffset+4 : file.DataOffset+4+fsp.HeaderV3Length])
```

Comment:
I'm not sure what to do - once I have the `uefi.File` object, which has this 4-byte header, what should I parse it with?

Reply:
(replied below)

Reply:
Does that uefi.File contain a uefi.Section? Is it a raw section?

Reply:
doesn't seem to be a section, if I am doing this right. I'm using `uefi.NewSection`, and I get the error

```
section size mismatch! Section has size 2572478, but buffer is 140 bytes big
```

However I couldn't figure out what `fileOrder` is, so I used 0 (which may be just plain wrong)

Reply:
My mistake. This works. Still unsure about file order though

Reply:
also, yes, it's a raw file, so sections are not parsed

Reply:
This is so stupid. Why did they use a raw file type? This way when you're parsing you have no idea whether to treat raw file contents as sections or not, because there's no signature and it could be garbage that you're reading as the size and type.